shortfin/python/shortfin_apps/llm/components/kvcache/page_pool.py

```diff
@@ -0,0 +1,159 @@
+    # 32: number of transformer blocks
+    # 2: one for k + one for v
+    # 16: tokens per page
+    # 8: head count (32 heads, but every 4 heads share the same kv buffer)
+    # 128: hidden dimension
+
+
+class PagePool:
+    """Page table based attention cache.
+
+    While internal to a model, the cache is organized with additional structure
+    per page, outside of the model, it is just a list of pages of a certain
+    element type and number of elements (all inner dims are flattened).
+
+    One page table is allocated per device in a fiber. Currently, this is a
+    dense allocation with committed memory but in the future, we may just
+    allocate the address space and lazily populate it with committed memory.
+
+    The cache is unique because usage of it can span fibers and concurrency
+    is implicitly managed at the block level (i.e. freshly acquired blocks
+    are assumed to be uninitialized and available immediately for use).
+
+    It is initialized with a discrete list of fiberd devices from a fiber but
+    cache usage can be done from any fiber which includes those devices.
+
+    In addition to supporting paged attention standalone, this also serves
+    as the array / buffer allocation layer for radix attention described in
+    `radix_tree.py`.
+    """
+
+    def __init__(self, *, devices: Sequence[sf.ScopedDevice], config: PagePoolConfig):
+        self._lock = threading.Lock()
+        self.devices = list(devices)
+        self.config = config
+        self.page_tables: list[sf.array.device_array] = []
+
+        # Setup accounting structs.
+        self.attn_page_entries = [
+            PageInfo(
+                index=i,
+                pool=self,
+                token_offset=0,
+                token_count=0,
+            )
+            for i in range(self.config.alloc_page_count)
+        ]
+
+        self.attn_page_free = list(self.attn_page_entries)
+
+        # Initialize a page table on each device.
+        page_table_shape = [
+            self.config.alloc_page_count,
+            self.config.paged_kv_block_size_elements,
+        ]
+        for device in devices:
+            logging.info(
+                "Allocating page table (shape=%r, dtype=%r, size=%s) on %r",
+                page_table_shape,
+                self.config.dtype,
+                human_size(config.dtype.compute_dense_nd_size(page_table_shape)),
+                device,
+            )
+            page_table = sf.array.device_array.for_device(
+                device, page_table_shape, self.config.dtype
+            )
+            self.page_tables.append(page_table)
+
+    def acquire_free_pages(self, count: int) -> list[PageInfo] | None:
+        with self._lock:
+            available = len(self.attn_page_free)
+            if count > available:
+                return None
+            return [self.attn_page_free.pop() for _ in range(count)]
```

Comment:
Kind of a nit, but might be relevant in cases where we're acquiring a large number of pages:

```python
pages = self.attn_page_free[:count]
self.attn_page_free = self.attn_page_free[count:]
return pages
```

Removes entire slice instead of individually popping and saves the overhead of creating a new list with list comprehension. I don't think we need to explicitly remove from the end of `self.attn_page_free`, so should be fine without the pop. If we do need to explicitly remove from end, we could negative index `count`. 

Does unfortunately turn a one-liner into 3, but this might be the case where it's worth it

Reply:
I think this is a good idea when we go larger

Check this out

https://colab.research.google.com/drive/1otgqrRgSkdc1d0hQ8Zxuq85UqeuGoCQ0?usp=sharing


    Size |      Pop Method |    Slice Method |      Ratio
-------------------------------------------------------
       1 |        0.000027 |        0.000165 |       0.17
      10 |        0.000031 |        0.000147 |       0.21
     100 |        0.000055 |        0.000149 |       0.37
    1000 |        0.000157 |        0.000153 |       1.02

Popping is much faster than slicing up to some 1000 pages at a time. But beyond that slicing gets slower very fast.

Reply:
hm but now that i think about it, at that point maybe we should just increase the page size.

Reply:
i think size 1 is just mostly dominated by the set-up so maybe we should toss that result!